ext/gmp/gmp.c

```diff
@@ -1284,37 +1284,23 @@ ZEND_FUNCTION(gmp_pow)
 
 	if (Z_TYPE_P(base_arg) == IS_LONG && Z_LVAL_P(base_arg) >= 0) {
 		INIT_GMP_RETVAL(gmpnum_result);
-		if (exp >= INT_MAX) {
-			mpz_t base_num, exp_num, mod;
-			mpz_init(base_num);
-			mpz_init(exp_num);
-			mpz_init(mod);
-			mpz_set_si(base_num, Z_LVAL_P(base_arg));
-			mpz_set_si(exp_num, exp);
-			mpz_set_ui(mod, UINT_MAX);
-			mpz_powm(gmpnum_result, base_num, exp_num, mod);
-			mpz_clear(mod);
-			mpz_clear(exp_num);
-			mpz_clear(base_num);
-		} else {
-			mpz_ui_pow_ui(gmpnum_result, Z_LVAL_P(base_arg), exp);
+		if ((log10(Z_LVAL_P(base_arg)) * exp) > (double)ULONG_MAX) {
```

Comment:
`ULONG_MAX` is only `2.147.483.647` on LLP64 (e.g. Windows 64bit). So that might constrain the range of allowed values too much. On the hand, on 32bit Windows that almost certainly can trigger OOM (e.g. `10 ** PHP_INT_MAX`).

Reply:
>  It might be better to declare mpir_ui gmpnum, and then go from there, but that might cause further issues down the line.

Note that the `mpir_ui` type does not exist in gmp (mpir lib ). `unsigned long int` might be appropriate however.

Reply:
Oh, of course; `mpir_ui` is an mpir type. `unsigned long int` is the same as `unsigned long`. We can probably use `zend_ulong`.

Reply:
Sorry yes I missed this `The problem is that mpz_get_ui() returns unsigned long long int (at least on Windows)`